File: src/maxx/enums.py
"""This module defines enumerations used by Malt."""

from __future__ import annotations

from enum import Enum


class Kind(str, Enum):
    """
    An enumeration representing different kinds of MATLAB code elements.
    This enumeration is a subclass of the Griffe `Kind` enumeration, and extends it with additional values.
    """

    FOLDER = "folder"
    """folders"""
    NAMESPACE = "namespace"
    """namespaces"""
    CLASS = "class"
    """Classes."""
    ENUMERATION = "class"
    """Class enumeration"""
    FUNCTION = "function"
    """Functions and methods."""
    SCRIPT = "script"
    """Scripts."""
    PROPERTY = "property"
    """Class properties."""
    ALIAS = "alias"
    """Aliases (imported objects)."""
    BUILTIN = "builtin"
    """Built-in objects."""


class ArgumentKind(str, Enum):
    """
    An enumeration representing different kinds of function arguments.

    Attributes:
        positional (str): Positional-only argument.
        optional (str): Optional argument.
        keyword_only (str): Keyword-only argument.
        varargin (str): Varargin argument.
    """

    positional_only = "positional-only"
    optional = "optional"
    keyword_only = "keyword-only"
    varargin = "varargin"


class AccessKind(str, Enum):
    """
    An enumeration representing different access levels for MATLAB code elements.

    Attributes:
        public (str): Represents public access level.
        protected (str): Represents protected access level.
        private (str): Represents private access level.
        immutable (str): Represents immutable access level.
    """

    public = "public"
    protected = "protected"
    private = "private"
    immutable = "immutable"

File: src/core/enums.rs
//! Core enumerations - pure Rust, no Python dependencies

use serde::{Deserialize, Serialize};

/// Different kinds of MATLAB code elements
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Kind {
    Folder,
    Namespace,
    Class,
    Enumeration,
    Function,
    Script,
    Property,
    Alias,
    Builtin,
}

impl Kind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Kind::Folder => "folder",
            Kind::Namespace => "namespace",
            Kind::Class => "class",
            Kind::Enumeration => "enumeration",
            Kind::Function => "function",
            Kind::Script => "script",
            Kind::Property => "property",
            Kind::Alias => "alias",
            Kind::Builtin => "builtin",
        }
    }
}

/// Different kinds of function arguments
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ArgumentKind {
    PositionalOnly,
    Optional,
    KeywordOnly,
    Varargin,
}

impl ArgumentKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            ArgumentKind::PositionalOnly => "positional-only",
            ArgumentKind::Optional => "optional",
            ArgumentKind::KeywordOnly => "keyword-only",
            ArgumentKind::Varargin => "varargin",
        }
    }
}

/// Access levels for MATLAB code elements
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum AccessKind {
    Public,
    Protected,
    Private,
    Immutable,
}

impl AccessKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            AccessKind::Public => "public",
            AccessKind::Protected => "protected",
            AccessKind::Private => "private",
            AccessKind::Immutable => "immutable",
        }
    }
}
